### PR TITLE
[multistage][bugfix] eliminate multiple exchanges

### LINKED ISSUE #11881: [multistage] multiple nested SEMI-join create wrong plan
running this in the colocated quickstart
```
explain plan for
WITH tmp2 AS (
  SELECT * FROM userGroups
  WHERE groupUUID NOT IN ('group-3', 'group-4')
),
tmp3 AS (
  SELECT *
  FROM userAttributes
  WHERE userUUID IN (SELECT userUUID FROM tmp2)
    AND totalTrips < 100
)
SELECT * FROM tmp3 WHERE userUUID IN (SELECT userUUID from tmp2)
```
produces
```
Execution Plan
LogicalJoin(condition=[=($3, $5)], joinType=[semi])
  PinotLogicalExchange(distribution=[hash[3]])      <-- THIS should be the right exchange
    PinotLogicalExchange(distribution=[hash[3]])      <-- THIS exchange is redundant
      LogicalJoin(condition=[=($3, $5)], joinType=[semi])
        LogicalProject(daysSinceFirstTrip=[$3], deviceOS=[$4], totalTrips=[$5], userUUID=[$6])
          LogicalFilter(condition=[<($5, 100)])
            LogicalTableScan(table=[[userAttributes]])
        PinotLogicalExchange(distribution=[broadcast], relExchangeType=[PIPELINE_BREAKER])
          LogicalProject(groupUUID=[$3], userUUID=[$4])
            LogicalFilter(condition=[AND(<>($3, 'group-3'), <>($3, 'group-4'))])
              LogicalTableScan(table=[[userGroups]])
  PinotLogicalExchange(distribution=[hash[1]])
    LogicalProject(groupUUID=[$3], userUUID=[$4])
      LogicalFilter(condition=[AND(<>($3, 'group-3'), <>($3, 'group-4'))])
        LogicalTableScan(table=[[userGroups]])
```

### PR DESCRIPTION
**Pull request by walterddr**
this fixes #11881 

this rule will be merged into a more sophisticated rule that will be created in #11831

### CODE REVIEW

File: pinot-query-planner/src/test/resources/queries/JoinPlans.json [hunk 1]
```diff
@@ -421,6 +421,28 @@
...
+          "\n    LogicalJoin(condition=[=($0, $7)], joinType=[semi])",
+          "\n      LogicalFilter(condition=[<($2, 100)])",
+          "\n        LogicalTableScan(table=[[b]])",
+          "\n      PinotLogicalExchange(distribution=[broadcast], relExchangeType=[PIPELINE_BREAKER])",
+          "\n        LogicalProject(col1=[$0], col2=[$1])",
+          "\n          LogicalFilter(condition=[AND(<>($1, 'bar'), <>($1, 'foo'))])",
+          "\n            LogicalTableScan(table=[[a]])",
+          "\n  PinotLogicalExchange(distribution=[hash[0]])",
+          "\n    LogicalProject(col1=[$0], col3=[$2])",
+          "\n      LogicalFilter(condition=[>($2, 10)])",
```
Comment (xiangfu0) [inline pinot-query-planner/src/test/resources/queries/JoinPlans.json:441]: This is interesting, I think visitFilter in ServerPlanRequestVisitor will override each other, so one predicate will be skipped.

Comment (walterddr) [reply pinot-query-planner/src/test/resources/queries/JoinPlans.json:441]: this is not unexpected. (we can potentially make it more smart). the problem is that
```
WITH tmp1
     AS (SELECT *
         FROM   a
         WHERE  col2 NOT IN ( 'foo', 'bar' )),
     tmp2
     AS (SELECT *
         FROM   b
         WHERE  col1 IN (SELECT col1
                         FROM   tmp1)
                AND col3 < 100)
SELECT *
FROM   tmp2
WHERE  col1 IN (SELECT col1
                FROM   tmp1  <-- this is the tmp1 modification into tmp1'
                WHERE  col3 > 10) 
```
at this stage Calcite's doesn't know if you want to do a table spool of tmp1 and make tmp1' derived from tmp1 --> thus col3 > 10 is not pushed down. 
only when we decided that table spool is not possible and then we decided to copy tmp1 and tmp1' into 2 separate sub queries, at that time i think the hep optimizer already get passed the filter merging phase. 

rewriting the query in this way
```
WITH tmp1
     AS (SELECT *
         FROM   a
         WHERE  col2 NOT IN ( 'foo', 'bar' )),
     tmp2
     AS (SELECT *
         FROM   b
         WHERE  col1 IN (SELECT col1
                         FROM   tmp1)
                AND col3 < 100),
     tmp3
     AS (SELECT *
         FROM   a
         WHERE  col2 NOT IN ( 'foo', 'bar' ) AND col3 > 10),
SELECT *
FROM   tmp2
WHERE  col1 IN (SELECT col1
                FROM   tmp3) 
```
produces no multi-filter plan

Comment (xiangfu0) [reply pinot-query-planner/src/test/resources/queries/JoinPlans.json:441]: If we cannot avoid this, then it means the ServerPlanRequestVisitor should handle the leaf query generation properly.

Comment (walterddr) [reply pinot-query-planner/src/test/resources/queries/JoinPlans.json:441]: this is a un-plannable sql before or after this bug fix PR. I would suggest differ the fix into a different PR (i will change the test to make sure it doesn't happen)

**Review by Jackie-Jiang — APPROVED** [review]

**Pull request merged**